### ISSUE TITLE
Topic names are not compared correctly

### ISSUE THREAD

**Issue by a1lu** [arobenko/mqtt-sn#8]
Hi,
I subscribe a topic `test` and get a confirmation `SUBACK` from the gateway. After that I try to publish to topic `test/test` but all my messages are send to `test`.
The problem is that the equality check in [BasicClient.h](https://github.com/arobenko/mqtt-sn/blob/develop/client/src/basic/BasicClient.h#L2250) is wrong.

A fix could be to check for equal length and for equal names.
```
 return elem.m_allocated && (elem.m_topic.size() == strlen(op->m_topic)) && (elem.m_topic == op->m_topic);
```

But at the end this seems to be a bug in COMMS. I did not check if this is fixed yet in a newer COMMS version and gets fixed here when we are compatible to new COMMS.
I can create a pull request on the weekend if you like.

**Comment by arobenko** [comment]
At this moment I use [ext_mqtt311](https://github.com/arobenko/mqtt-sn/tree/ext_mqtt311) branch (instead of develop), which uses latest COMMS library as well as [cc.mqtt311.generated](https://github.com/arobenko/cc.mqtt311.generated) (instead of [mqtt](https://github.com/arobenko/mqtt)) for MQTT protocol definition (used by gateway).

I think it's getting close to be officially released (don't have #7 to be merged yet). Could you please try it and say whether it works for you?

**Comment by arobenko** [comment]
I can't reproduce reported problem with "ext_mqtt311" branch.

**Comment by arobenko** [comment]
Just FYI "ext_mqtt311" was merged to develop

**Comment by a1lu** [comment]
I can still reproduce this with current `develop` branch,

**Comment by arobenko** [comment]
Can you please give me some more details on how to reproduce? What gw and
what pub sub utilities you are using, exactly what topics? Are they
predefined or regular, etc?

**Comment by a1lu** [comment]
I could not reproduce this on a blank pull from develop, but if I integrate it into my project. I noticed the difference in the implementation of `operator==`.
The blank branch uses 
`std::operator==`. 
In my integration I use the config
`set(MQTTSN_CUSTOM_CLIENT_TOPIC_NAME_STATIC_STORAGE_SIZE 64)`,
this leads to
 `comms::util::StaticString<64ul, char>::operator==`.
 Implementation:
```
bool operator==(const TChar* str) const
    {
        for (auto idx = 0U; idx < size(); ++idx) {
            if (*str == Ends) {
                return false;
            }

            auto ch = vec_[idx];
            if (*str != ch) {
                return false;
            }

            ++str;
        }

        return true;
    }
```

`str = test/abc`
`this->vec_ = test`
The loop ends after consuming `test` and returns `true`

**Comment by arobenko** [comment]
Thanks, good find and analysis. I'll introduce fix to StaticString in the
comms library early next week.

Last 'return true' needs to be replaced with string termination check.

**Comment by arobenko** [comment]
Just FYI: At this moment "develop" branch of [comms_champion](https://github.com/arobenko/comms_champion) project contains fixes to StaticString. I haven't released the fixes yet and as the result haven't integrated into this project yet.

**Comment by arobenko** [comment]
The "develop" branch of this project in conjunction with develop branch of comms_champion project should have the problem fixed. I'll prepare the official releases soon.

Also note, that I introduced new cmake variable **CC_MQTTSN_DEFAULT_CLIENT_CONFIG_FILE** (similar to **CC_MQTTSN_CUSTOM_CLIENT_CONFIG_FILES**). It can be used to provide custom configuration for the **default** library, used by the pub / sub utilities. It can be used to easily test your embedded configuration on the development PC.

**Comment by arobenko** [comment]
Should be fixed in [v0.9](https://github.com/arobenko/mqtt-sn/releases/tag/v0.9) release, please reopen if problem persists.
